File: backend/app/plans.py
"""Plan tiers and their capabilities — canonical copy of docs/PLANS.md.

Keep this in sync with `frontend/src/lib/plans.ts` and `docs/PLANS.md`.
"""

from __future__ import annotations

from enum import Enum

from .config import settings


class PlanTier(str, Enum):
    ANONYMOUS = "anonymous"
    PRO = "pro"
    PREMIUM = "premium"


# Paid tiers that require an account. Registration always lands here.
ACCOUNT_TIERS = (PlanTier.PRO, PlanTier.PREMIUM)

# Capability flags. `True` / `False`, or a scalar limit.
FEATURES: dict[str, dict[PlanTier, object]] = {
    "unicode_analysis": {PlanTier.ANONYMOUS: True, PlanTier.PRO: True, PlanTier.PREMIUM: True},
    "basic_metadata": {PlanTier.ANONYMOUS: True, PlanTier.PRO: True, PlanTier.PREMIUM: True},
    "full_metadata": {PlanTier.ANONYMOUS: False, PlanTier.PRO: True, PlanTier.PREMIUM: True},
    "c2pa": {PlanTier.ANONYMOUS: True, PlanTier.PRO: True, PlanTier.PREMIUM: True},
    "statistical_analysis": {PlanTier.ANONYMOUS: False, PlanTier.PRO: True, PlanTier.PREMIUM: True},
    "fingerprint_matching": {PlanTier.ANONYMOUS: False, PlanTier.PRO: True, PlanTier.PREMIUM: True},
    "detailed_report": {PlanTier.ANONYMOUS: False, PlanTier.PRO: False, PlanTier.PREMIUM: True},
    "server_history": {PlanTier.ANONYMOUS: False, PlanTier.PRO: True, PlanTier.PREMIUM: True},
    "clean_text": {PlanTier.ANONYMOUS: True, PlanTier.PRO: True, PlanTier.PREMIUM: True},
    "clean_image_metadata": {PlanTier.ANONYMOUS: True, PlanTier.PRO: True, PlanTier.PREMIUM: True},
    "clean_documents": {PlanTier.ANONYMOUS: False, PlanTier.PRO: True, PlanTier.PREMIUM: True},
    "clean_batch": {PlanTier.ANONYMOUS: False, PlanTier.PRO: False, PlanTier.PREMIUM: True},
    "report_export": {PlanTier.ANONYMOUS: False, PlanTier.PRO: True, PlanTier.PREMIUM: True},
    "batch_analysis": {PlanTier.ANONYMOUS: False, PlanTier.PRO: False, PlanTier.PREMIUM: True},
    "api_access": {PlanTier.ANONYMOUS: False, PlanTier.PRO: False, PlanTier.PREMIUM: True},
}

CONTENT_TYPES: dict[PlanTier, list[str]] = {
    PlanTier.ANONYMOUS: ["text", "code", "image"],
    PlanTier.PRO: ["text", "code", "image", "pdf", "docx", "audio"],
    PlanTier.PREMIUM: ["text", "code", "image", "pdf", "docx", "audio", "video"],
}


def can(plan: PlanTier, feature: str) -> bool:
    return bool(FEATURES.get(feature, {}).get(plan, False))


def scan_limit(plan: PlanTier) -> int | None:
    """Scans allowed per window. ``None`` means unlimited."""
    return {
        PlanTier.ANONYMOUS: settings.anon_scan_limit,
        PlanTier.PRO: settings.pro_scan_limit,
        PlanTier.PREMIUM: None,
    }[plan]


def window_hours(plan: PlanTier) -> int:
    return {
        PlanTier.ANONYMOUS: settings.anon_window_hours,
        PlanTier.PRO: settings.pro_window_hours,
        PlanTier.PREMIUM: settings.pro_window_hours,
    }[plan]


def max_file_bytes(plan: PlanTier) -> int:
    return {
        PlanTier.ANONYMOUS: settings.anon_max_file_bytes,
        PlanTier.PRO: settings.pro_max_file_bytes,
        PlanTier.PREMIUM: settings.premium_max_file_bytes,
    }[plan]


def describe(plan: PlanTier) -> dict:
    """Full capability snapshot for a tier — consumed by the frontend (camelCase)."""
    return {
        "tier": plan.value,
        "requiresAccount": plan in ACCOUNT_TIERS,
        "scanLimit": scan_limit(plan),
        "windowHours": window_hours(plan),
        "maxFileBytes": max_file_bytes(plan),
        "contentTypes": CONTENT_TYPES[plan],
        "features": {name: can(plan, name) for name in FEATURES},
    }


def describe_all() -> list[dict]:
    return [describe(p) for p in PlanTier]

File: frontend/src/lib/plans.ts
/**
 * Plan capability matrix — mirror of backend/app/plans.py and docs/PLANS.md.
 * The backend is authoritative (GET /meta/plans); this copy powers instant,
 * offline UI gating.
 */
import type { PlanTier } from '@/types/user';

export type Feature =
  | 'unicode_analysis'
  | 'basic_metadata'
  | 'full_metadata'
  | 'c2pa'
  | 'statistical_analysis'
  | 'fingerprint_matching'
  | 'detailed_report'
  | 'server_history'
  | 'clean_text'
  | 'clean_image_metadata'
  | 'clean_documents'
  | 'clean_batch'
  | 'report_export'
  | 'batch_analysis'
  | 'api_access';

export interface PlanCapabilities {
  tier: PlanTier;
  requiresAccount: boolean;
  scanLimit: number | null;
  windowHours: number;
  maxFileBytes: number;
  contentTypes: string[];
  features: Record<Feature, boolean>;
}

const F = (list: Feature[]): Record<Feature, boolean> => {
  const all: Feature[] = [
    'unicode_analysis', 'basic_metadata', 'full_metadata', 'c2pa',
    'statistical_analysis', 'fingerprint_matching', 'detailed_report',
    'server_history', 'clean_text', 'clean_image_metadata', 'clean_documents',
    'clean_batch', 'report_export', 'batch_analysis', 'api_access',
  ];
  return Object.fromEntries(all.map((f) => [f, list.includes(f)])) as Record<Feature, boolean>;
};

export const PLANS: Record<PlanTier, PlanCapabilities> = {
  anonymous: {
    tier: 'anonymous',
    requiresAccount: false,
    scanLimit: 5,
    windowHours: 48,
    maxFileBytes: 2 * 1024 * 1024,
    contentTypes: ['text', 'code', 'image'],
    features: F(['unicode_analysis', 'basic_metadata', 'clean_text', 'clean_image_metadata']),
  },
  pro: {
    tier: 'pro',
    requiresAccount: true,
    scanLimit: 300,
    windowHours: 24,
    maxFileBytes: 50 * 1024 * 1024,
    contentTypes: ['text', 'code', 'image', 'pdf', 'docx', 'audio'],
    features: F([
      'unicode_analysis', 'basic_metadata', 'full_metadata', 'c2pa',
      'statistical_analysis', 'fingerprint_matching', 'server_history',
      'clean_text', 'clean_image_metadata', 'clean_documents', 'report_export',
    ]),
  },
  premium: {
    tier: 'premium',
    requiresAccount: true,
    scanLimit: null,
    windowHours: 24,
    maxFileBytes: 200 * 1024 * 1024,
    contentTypes: ['text', 'code', 'image', 'pdf', 'docx', 'audio', 'video'],
    features: F([
      'unicode_analysis', 'basic_metadata', 'full_metadata', 'c2pa',
      'statistical_analysis', 'fingerprint_matching', 'detailed_report',
      'server_history', 'clean_text', 'clean_image_metadata', 'clean_documents',
      'clean_batch', 'report_export', 'batch_analysis', 'api_access',
    ]),
  },
};

/** Live matrix, replaced once GET /meta/plans resolves. */
let matrix: Record<PlanTier, PlanCapabilities> = PLANS;

export function setPlanMatrix(next: Partial<Record<PlanTier, PlanCapabilities>>) {
  matrix = { ...matrix, ...next };
}

export function planFor(tier: PlanTier): PlanCapabilities {
  return matrix[tier] ?? PLANS[tier];
}

export function can(tier: PlanTier, feature: Feature): boolean {
  return Boolean(planFor(tier).features[feature]);
}

export const PLAN_LABELS: Record<PlanTier, string> = {
  anonymous: 'Free',
  pro: 'Pro',
  premium: 'Premium',
};

export function formatBytes(bytes: number): string {
  if (bytes >= 1024 * 1024) return `${Math.round(bytes / (1024 * 1024))} MB`;
  if (bytes >= 1024) return `${Math.round(bytes / 1024)} KB`;
  return `${bytes} B`;
}
